libmamba/src/core/satisfiability_error.cpp

```diff
@@ -0,0 +1,433 @@
+                case SOLVER_RULE_PKG_RECOMMENDS:
+                case SOLVER_RULE_RECOMMENDS:
+                case SOLVER_RULE_UNKNOWN:
+                    return std::nullopt;
+                case SOLVER_RULE_PKG_SELF_CONFLICT:
+                case SOLVER_RULE_PKG_OBSOLETES:
+                case SOLVER_RULE_PKG_IMPLICIT_OBSOLETES:
+                case SOLVER_RULE_PKG_INSTALLED_OBSOLETES:
+                case SOLVER_RULE_YUMOBS:
+                case SOLVER_RULE_DISTUPGRADE:
+                    return std::nullopt;
+            }
+            return std::nullopt;
+        }
+
+        class ProblemsGraphCreator
+        {
+        public:
+            using SolvId = Id;  // Unscoped from libsolv
+
+            using graph_t = ProblemsGraph::graph_t;
+            using ProblemType = ProblemsGraph::ProblemType;
+            using RootNode = ProblemsGraph::RootNode;
+            using PackageNode = ProblemsGraph::PackageNode;
+            using UnresolvedDependencyNode = ProblemsGraph::UnresolvedDependencyNode;
+            using node_t = ProblemsGraph::node_t;
+            using node_id = ProblemsGraph::node_id;
+            using ConstraintEdge = ProblemsGraph::ConstraintEdge;
+            using RequireEdge = ProblemsGraph::RequireEdge;
+            using edge_t = ProblemsGraph::edge_t;
+            using conflict_map = ProblemsGraph::conflict_map;
+
+
+            ProblemsGraphCreator(MSolver const& solver, MPool const& pool)
+                : m_solver{ solver }
+                , m_pool{ pool }
+            {
+                m_root_node = m_graph.add_node(RootNode());
+                parse_problems();
+            }
+
+            operator ProblemsGraph() &&
+            {
+                return { std::move(m_graph), std::move(m_conflicts), m_root_node };
+            }
+
+
+        private:
+            MSolver const& m_solver;
+            MPool const& m_pool;
+            graph_t m_graph;
+            conflict_map m_conflicts;
+            std::map<SolvId, node_id> m_solv2node;
+            node_id m_root_node;
+
+
+            /**
+             * Add a node if it is not already present and return the node id.
+             */
+            auto ensure_solvable(SolvId solv_id, node_t&& pkg_info) -> node_id;
+            auto ensure_solvable(SolvId solv_id,
+                                 PackageInfo&& pkg_info,
+                                 std::optional<ProblemType> pb_type) -> node_id;
+            auto ensure_solvable(SolvId solv_id,
+                                 std::string&& dep,
+                                 std::optional<ProblemType> pb_type) -> node_id;
+
+            auto add_conflict(node_id n1, node_id n2) -> void;
+            auto add_expanded_deps_edges(node_id from_id, SolvId dep_id, edge_t const& edge)
+                -> void;
+            auto parse_problems() -> void;
+        };
+
+        auto ProblemsGraphCreator::ensure_solvable(SolvId solv_id, node_t&& node) -> node_id
```

Comment:
the naming is confusing to me, I would use something like get_or_create_solvable but up to you